twilio/twilio.d.ts

```diff
@@ -0,0 +1,1408 @@
+
+    get: RestMethod;
+
+    daily: UsageRecordRange;
+    monthly: UsageRecordRange;
+    yearly: UsageRecordRange;
+    allTime: UsageRecordRange;
+    today: UsageRecordRange;
+    yesterday: UsageRecordRange;
+    thisMonth: UsageRecordRange;
+    lastMonth: UsageRecordRange;
+  }
+
+  /// resources/ip_messaging/Credentials.js
+  export interface CredentialInstance {
+    get: RestMethod;
+    post: RestMethod;
+    delete: RestMethod;
+    update: RestMethod;
+  }
+
+  export interface CredentialResource {
+    (resourceSid: string): CredentialInstance;
+
+    get: RestMethod;
+    post: RestMethod;
+    create: RestMethod;
+    list: RestMethod;
+  }
+
+  /// resources/ip_messaging/Services.js
+  export interface ServiceUserInstance {
+    get: RestMethod;
+    post: RestMethod;
+    delete: RestMethod;
+    update: RestMethod;
+  }
+
+  export interface ServiceUserResource {
+    (resourceSid: string): ServiceUserInstance;
+
+    get: RestMethod;
+    post: RestMethod;
+    create: RestMethod;
+    list: RestMethod;
+  }
+
+  export interface ServiceRoleInstance {
+    get: RestMethod;
+  }
+
+  export interface ServiceRoleResource {
+    (resourceSid: string): ServiceRoleInstance;
+
+    get: RestMethod;
+    list: RestMethod;
+  }
+
+  export interface ServiceChannelMessageInstance {
+    get: RestMethod;
+    post: RestMethod;
+    delete: RestMethod;
+    update: RestMethod;
+  }
+
+  export interface ServiceChannelMessageResource {
+    (resourceSid: string): ServiceChannelMessageInstance;
+
+    get: RestMethod;
+    post: RestMethod;
+    create: RestMethod;
+    list: RestMethod;
+  }
+
+  export interface ServiceChannelMemberInstance {
+    get: RestMethod;
+    post: RestMethod;
+    delete: RestMethod;
+    update: RestMethod;
+  }
+
+  export interface ServiceChannelMemberResource {
+    (resourceSid: string): ServiceChannelMemberInstance;
+
+    get: RestMethod;
+    post: RestMethod;
+    create: RestMethod;
+    list: RestMethod;
+  }
+
+  export interface ServiceChannelInstance {
+    get: RestMethod;
+    post: RestMethod;
+    delete: RestMethod;
+    update: RestMethod;
+
+    messages: ServiceChannelMessageResource;
+    members: ServiceChannelMemberResource;
+  }
+
+  export interface ServiceChannelResource {
+    (resourceSid: string): ServiceChannelInstance;
+
+    get: RestMethod;
+    post: RestMethod;
+    create: RestMethod;
+    list: RestMethod;
+  }
+
+  export interface ServiceInstance {
+    get: RestMethod;
+    post: RestMethod;
+    delete: RestMethod;
+    update: RestMethod;
+
+    users: ServiceUserResource;
+    roles: ServiceRoleResource;
+    channels: ServiceChannelResource;
+  }
+
+  export interface ServiceResource {
+    (resourceSid: string): ServiceInstance;
+
+    get: RestMethod;
+    post: RestMethod;
+    create: RestMethod;
+    list: RestMethod;
+  }
+
+  /// resources/lookups/PhoneNumbers.js
+  export interface PhoneNumberInstance {
+    get: RestMethod;
```

Comment:
a lot of these classes are structurally identical (PhoneNumberInstance, EventInstance, AlertInstance, ...) or very close (AlertResource, EventResource, CountryResource, ...). Is it really worthwhile to have so many different ones? The compiler will treat them all the same &mdash; you'll be able to assign an AlertInstance to a PhoneNumberInstance property, for example.